services/pull/pull.go

```diff
@@ -38,6 +38,114 @@ func NewPullRequest(repo *models.Repository, pull *models.Issue, labelIDs []int6
 	return nil
 }
 
+// ChangeTargetBranch changes the target branch of this pull request, as the given user.
+func ChangeTargetBranch(pr *models.PullRequest, doer *models.User, targetBranch string) (err error) {
+	// Current target branch is already the same
+	if pr.BaseBranch == targetBranch {
+		return nil
+	}
+
+	if pr.Issue.IsClosed {
+		return models.ErrIssueIsClosed{
+			ID:     pr.Issue.ID,
+			RepoID: pr.Issue.RepoID,
+			Index:  pr.Issue.Index,
+		}
+	}
+
+	if pr.HasMerged {
+		return models.ErrPullRequestHasMerged{
+			ID:         pr.ID,
+			IssueID:    pr.Index,
+			HeadRepoID: pr.HeadRepoID,
+			BaseRepoID: pr.BaseRepoID,
+			HeadBranch: pr.HeadBranch,
+			BaseBranch: pr.BaseBranch,
+		}
+	}
+
+	// Check if branches are equal
+	if err = pr.GetBaseRepo(); err != nil {
+		return err
+	}
+	baseGitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
+	if err != nil {
+		return err
+	}
+	baseCommit, err := baseGitRepo.GetBranchCommit(targetBranch)
+	if err != nil {
+		return err
+	}
+
+	if err = pr.GetHeadRepo(); err != nil {
+		return err
+	}
+	headGitRepo, err := git.OpenRepository(pr.HeadRepo.RepoPath())
+	if err != nil {
+		return err
+	}
+	headCommit, err := headGitRepo.GetBranchCommit(pr.HeadBranch)
+	if err != nil {
+		return err
+	}
+	branchesEqual, err := baseCommit.HasPreviousCommit(headCommit.ID)
+	if err != nil {
+		return err
+	}
+	if branchesEqual {
+		return models.ErrBranchesEqual{
+			HeadBranchName: pr.HeadBranch,
+			BaseBranchName: targetBranch,
+		}
+	}
+
+	// Check if pull request for the new target branch already exists
+	existingPr, err := models.GetUnmergedPullRequest(pr.HeadRepoID, pr.BaseRepoID, pr.HeadBranch, targetBranch)
+	if existingPr != nil {
+		return models.ErrPullRequestAlreadyExists{
+			ID:         existingPr.ID,
+			IssueID:    existingPr.Index,
+			HeadRepoID: existingPr.HeadRepoID,
+			BaseRepoID: existingPr.BaseRepoID,
+			HeadBranch: existingPr.HeadBranch,
+			BaseBranch: existingPr.BaseBranch,
+		}
+	}
+	if err != nil && !models.IsErrPullRequestNotExist(err) {
+		return err
+	}
+
+	// Set new target branch
+	oldBranch := pr.BaseBranch
+	pr.BaseBranch = targetBranch
+
+	if err := TestPatch(pr); err != nil {
+		return err
+	}
+
+	if err := pr.UpdateCols("base_branch"); err != nil {
```

Comment:
I think we should update the database on a transaction. This should be merged with `checkAndUpdateStatus`.

Reply:
Hmm, I get see that one transaction is better than two. I just didn't want to copy code that is already in a function. Maybe I could add an optional param to `checkAndUpdateStatus` which allows updating additional columns with that transaction?

Reply:
> Hmm, I get see that one transaction is better than two. I just didn't want to copy code that is already in a function. Maybe I could add a parameter to `checkAndUpdateStatus` which allows updating additional columns with that transaction?

Didn't work as expected. I'll just merge the contents of that function with my transaction...